Add support for background preferred_environment in Web Extensions.

xeenon opened the pull request:
#### 6d27093d068a866cc0f2948e561882d1f2d8039b
<pre>
Add support for background preferred_environment in Web Extensions.
<a href="https://webkit.org/b/272244">https://webkit.org/b/272244</a>
<a href="https://rdar.apple.com/problem/125988233">rdar://problem/125988233</a>

Reviewed by Brian Weinstein.

Add support for extensions declaring a `preferred_environment` in the `background` entry.
The `preferred_environment` key can be a string or an array of strings. When an array is
used, the browser will prefer the first one it supports.

    &quot;background&quot;: {
        &quot;preferred_environment&quot;: [ &quot;service_worker&quot;, &quot;document&quot; ],
        &quot;scripts&quot;: [ &quot;script1.js&quot;, &quot;script2.js&quot; ]
    }

If no environments are specified or supported, then we fallback to the previous handling of
the `scripts`, `page` or `service_worker` keys (in that order) for backwards compatibility.

This allows cross-browser extensions to have a single manifest that works in multiple
browsers. Chrome plans to only support service workers going forward, but Safari and
Firefox plan to support pages and service workers, and extension might prefer a document
environment over a service worker in Safari and Firefox.

This also adds support for a `_generated_service_worker.js` script that works with normal
scripts or modules based on the `scripts` and `type` keys if `preferred_environment`
is `service_worker`.

WECG discussion here: <a href="https://github.com/w3c/webextensions/issues/282">https://github.com/w3c/webextensions/issues/282</a>

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::resourceStringForPath): Support _generated_service_worker.js.
(WebKit::WebExtension::resourceDataForPath): Ditto.
(WebKit::WebExtension::backgroundContentUsesModules): Use renamed m_backgroundContentUsesModules.
(WebKit::WebExtension::backgroundContentIsServiceWorker): Use new m_backgroundContentEnvironment.
(WebKit::WebExtension::backgroundContentPath): Support _generated_service_worker.js.
(WebKit::WebExtension::generatedBackgroundContent): Generate the service worker script.
(WebKit::WebExtension::populateBackgroundPropertiesIfNeeded): Handle preferred_environment.
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Added m_backgroundContentEnvironment and renamed
m_backgroundPageUsesModules to m_backgroundContentUsesModules.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST(WKWebExtension, BackgroundPreferredEnvironmentParsing)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithServiceWorkerPreferredEnvironment)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithPageDocumentPreferredEnvironment)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithScriptsDocumentPreferredEnvironment)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithMultipleDocumentModuleScripts)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithMultipleServiceWorkerScripts)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithMultipleServiceWorkerModuleScripts)): Added.

Canonical link: <a href="https://commits.webkit.org/277228@main">https://commits.webkit.org/277228@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b17e69c05a88155b65d3a6ea8b45eca4913e9609

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47055 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49693 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49738 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43102 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49362 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31426 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23691 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47636 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/23681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/19633 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/21051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5100 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51614 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/22077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23357 "Failed to checkout and rebase branch from PR 26915") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44612 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/24136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6604 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/23069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1301,20 +1312,31 @@ static _WKWebExtensionError toAPI(WebExtension::Error error)
 
-    if (m_backgroundServiceWorkerPath)
+    if (m_backgroundServiceWorkerPath || m_backgroundPagePath)
         return nil;
```
Does this mean that the content has already been generated?

xeenon replied:
It means we shouldn't generate content since we have paths. These could be ASSERTs instead, but an extension can still give here if they try to load the generated page / script path.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1341,26 +1363,85 @@ static _WKWebExtensionError toAPI(WebExtension::Error error)
-        m_backgroundServiceWorkerPath = nil;
+    static auto *supportedEnvironments = [NSOrderedSet orderedSetWithObjects:backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey, nil];
+
+    NSOrderedSet *preferedEnvironments;
```
Nit: `preferredEnvironments`.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1341,26 +1363,85 @@ static _WKWebExtensionError toAPI(WebExtension::Error error)
+        if ([supportedEnvironments containsObject:environment])
+            preferedEnvironments = [NSOrderedSet orderedSetWithObject:environment];
+    } else if (auto *environments = objectForKey<NSArray>(backgroundManifestDictionary, backgroundPreferredEnvironmentManifestKey, true, NSString.class)) {
+        auto *filteredEnvironments = filterObjects(environments, ^bool(NSNumber *index, NSString *environment) {
```
Nit: Do we need the `index` variable to be named?

xeenon committed:
Add support for background preferred_environment in Web Extensions.
https://webkit.org/b/272244
rdar://problem/125988233

Reviewed by Brian Weinstein.

Add support for extensions declaring a `preferred_environment` in the `background` entry.
The `preferred_environment` key can be a string or an array of strings. When an array is
used, the browser will prefer the first one it supports.

    "background": {
        "preferred_environment": [ "service_worker", "document" ],
        "scripts": [ "script1.js", "script2.js" ]
    }

If no environments are specified or supported, then we fallback to the previous handling of
the `scripts`, `page` or `service_worker` keys (in that order) for backwards compatibility.

This allows cross-browser extensions to have a single manifest that works in multiple
browsers. Chrome plans to only support service workers going forward, but Safari and
Firefox plan to support pages and service workers, and extension might prefer a document
environment over a service worker in Safari and Firefox.

This also adds support for a `_generated_service_worker.js` script that works with normal
scripts or modules based on the `scripts` and `type` keys if `preferred_environment`
is `service_worker`.

WECG discussion here: https://github.com/w3c/webextensions/issues/282

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::resourceStringForPath): Support _generated_service_worker.js.
(WebKit::WebExtension::resourceDataForPath): Ditto.
(WebKit::WebExtension::backgroundContentUsesModules): Use renamed m_backgroundContentUsesModules.
(WebKit::WebExtension::backgroundContentIsServiceWorker): Use new m_backgroundContentEnvironment.
(WebKit::WebExtension::backgroundContentPath): Support _generated_service_worker.js.
(WebKit::WebExtension::generatedBackgroundContent): Generate the service worker script.
(WebKit::WebExtension::populateBackgroundPropertiesIfNeeded): Handle preferred_environment.
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Added m_backgroundContentEnvironment and renamed
m_backgroundPageUsesModules to m_backgroundContentUsesModules.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST(WKWebExtension, BackgroundPreferredEnvironmentParsing)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithServiceWorkerPreferredEnvironment)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithPageDocumentPreferredEnvironment)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithScriptsDocumentPreferredEnvironment)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithMultipleDocumentModuleScripts)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithMultipleServiceWorkerScripts)): Added.
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundWithMultipleServiceWorkerModuleScripts)): Added.

Canonical link: https://commits.webkit.org/277228@main
